File: users-service/src/main/java/net/plshark/users/auth/service/Ecdsa256AlgorithmBuilder.java
package net.plshark.users.auth.service;

import java.io.FileInputStream;
import java.io.IOException;
import java.security.GeneralSecurityException;
import java.security.KeyStore;
import java.security.interfaces.ECPrivateKey;
import java.security.interfaces.ECPublicKey;
import com.auth0.jwt.algorithms.Algorithm;
import net.plshark.users.auth.AuthProperties;
import org.springframework.stereotype.Component;

/**
 * Builder for the ECDSA256 algorithm
 */
@Component
public class Ecdsa256AlgorithmBuilder implements AlgorithmBuilder {

    public static final String ECDSA256 = "ecdsa256";

    @Override
    public Algorithm build(AuthProperties authProperties) throws IOException, GeneralSecurityException {
        Algorithm algorithm = null;
        String name = authProperties.getAlgorithm();
        if (ECDSA256.equals(name)) {
            if (authProperties.getKeystore() == null)
                throw new IllegalStateException("keystore must be set when using the ECDSA256 algorithm");
            if (authProperties.getKey() == null)
                throw new IllegalStateException("key must be set when using the ECDSA256 algorithm");

            KeyStore keyStore = KeyStore.getInstance(authProperties.getKeystore().getType());
            try (FileInputStream stream = new FileInputStream(authProperties.getKeystore().getLocation())) {
                keyStore.load(stream, authProperties.getKeystore().getPassword().toCharArray());
            }
            ECPrivateKey privateKey = (ECPrivateKey) keyStore.getKey(authProperties.getKey().getAlias(),
                    authProperties.getKey().getPassword().toCharArray());
            ECPublicKey publicKey = (ECPublicKey) keyStore.getCertificate(authProperties.getKey().getAlias()).getPublicKey();
            algorithm = Algorithm.ECDSA256(publicKey, privateKey);
        }
        return algorithm;
    }
}

File: user-error-server/src/main/kotlin/net/plshark/users/auth/service/Ecdsa256AlgorithmBuilder.kt
package net.plshark.users.auth.service

import com.auth0.jwt.algorithms.Algorithm
import net.plshark.users.auth.AuthProperties
import org.springframework.stereotype.Component
import java.io.FileInputStream
import java.io.IOException
import java.security.GeneralSecurityException
import java.security.KeyStore
import java.security.interfaces.ECPrivateKey
import java.security.interfaces.ECPublicKey

/**
 * Builder for the ECDSA256 algorithm
 */
@Component
class Ecdsa256AlgorithmBuilder : AlgorithmBuilder {

    @Throws(IOException::class, GeneralSecurityException::class)
    override fun build(authProperties: AuthProperties): Algorithm? {
        var algorithm: Algorithm? = null
        val name: String = authProperties.algorithm
        if (ECDSA256 == name) {
            checkNotNull(authProperties.keystore) { "keystore must be set when using the ECDSA256 algorithm" }
            checkNotNull(authProperties.key) { "key must be set when using the ECDSA256 algorithm" }

            val keyStore = KeyStore.getInstance(authProperties.keystore.type)
            FileInputStream(authProperties.keystore.location).use { stream ->
                keyStore.load(stream, authProperties.keystore.password.toCharArray())
            }
            val privateKey = keyStore.getKey(authProperties.key.alias, authProperties.key.password.toCharArray()) as
                ECPrivateKey
            val publicKey = keyStore.getCertificate(authProperties.key.alias).publicKey as ECPublicKey
            algorithm = Algorithm.ECDSA256(publicKey, privateKey)
        }
        return algorithm
    }

    companion object {
        const val ECDSA256 = "ecdsa256"
    }
}
